### PR TITLE
ffmpeg, ffmpeg_5: add patches for several CVEs

### PR DESCRIPTION
**Pull request by risicle**
## Description of changes

Most of these CVEs had PoCs, so (after struggling to build an ffmpeg with asan) I tested the PoCs to see which appeared to affect 6.1 and 5.1. 7.x fixes all of these.

CVE-2023-49501: 6.1.1 affected, patched. 5.1.4 unaffected?
CVE-2023-49502: 6.1.1 affected, patched. 5.1.4 affected, needed an adjusted patch.
CVE-2023-50007: 6.1.1 affected, patched. 5.1.4 unaffected?
CVE-2023-50008: both affected, patched.
CVE-2023-50009: 6.1.1 already includes fix, 5.1.4 affected but patch won't apply... TODO?
CVE-2023-50010: 6.1.1 already includes fix, 5.1.4 unaffected?
CVE-2023-51791: 6.1.1 already includes fix, 5.1.4 unaffected?
CVE-2023-51793: 6.1.1 already includes fix, 5.1.4 affected, patched
CVE-2023-51795: 6.1.1 already includes fix, 5.1.4 unaffected?
CVE-2023-51796: 6.1.1 already includes fix, 5.1.4 affected, patched

"unaffected" have a question mark because it's hard to be definite about this. Just because the exact supplied PoC didn't happen to trip it up in this version, doesn't mean a very similar vulnerability couldn't potentially be lurking.

To be able to include the in-tree patch, I needed to refactor the patch section to remove the `fetchpatch2` mapping.. repeating `fetchpatch2` for each entry really isn't too bad and the extra flexibility it gives us is needed.

Also among this batch were CVE-2023-51797 & CVE-2023-51798 which I couldn't reproduce or find a fix for.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/ffmpeg/generic.nix [hunk 1]
```diff
@@ -395,22 +395,22 @@ stdenv.mkDerivation (finalAttrs: {
       --replace /usr/local/lib/frei0r-1 ${frei0r}/lib/frei0r-1
   '';
 
-  patches = map (patch: fetchpatch2 patch) ([ ]
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/ffmpeg/generic.nix:398]:
```
  patches = [
    ...
  ] ++ map (patch: fetchpatch2 patch) ([ ]
```

Wouldn't this work?

Comment (risicle) [reply pkgs/development/libraries/ffmpeg/generic.nix:398]: But then I'd need to repeat the whole `optionals (lib.versionAtLeast version "5" && lib.versionOlder version "6")` stuff for the non-fetchpatched patches

Comment (Atemu) [reply pkgs/development/libraries/ffmpeg/generic.nix:398]: You could do a hack like

```Nix
map (patch: if builtins.isPath patch then patch else fetchpatch2 patch)
```

but I don't have a strong opinion on this.

**Review by Atemu — COMMENTED** [review]
Thanks a whole bunch for doing this!

Do you know whether ffmpeg will release new patch versions for these major versions soon? That'd obviate requiring all these patches.

File: pkgs/development/libraries/ffmpeg/generic.nix (hunk 1 above)
Comment (risicle) [reply pkgs/development/libraries/ffmpeg/generic.nix:398]: I just don't think it's a hugely useful piece of DRY in the first place, don't think we need the added complexity.

**Comment by risicle** [comment]
> Do you know whether ffmpeg will release new patch versions for these major versions soon? That'd obviate requiring all these patches.

I think it's unpredictable, but the patches I've added are from the end of 2023, so they had plenty of time to be backported if they were going to be.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 318912` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>ffmpeg</li>
    <li>ffmpeg.bin (ffmpeg.bin.bin ,ffmpeg.bin.data ,ffmpeg.bin.dev ,ffmpeg.bin.doc ,ffmpeg.bin.lib ,ffmpeg.bin.man)</li>
    <li>ffmpeg.data (ffmpeg.data.bin ,ffmpeg.data.data ,ffmpeg.data.dev ,ffmpeg.data.doc ,ffmpeg.data.lib ,ffmpeg.data.man)</li>
    <li>ffmpeg.dev (ffmpeg.dev.bin ,ffmpeg.dev.data ,ffmpeg.dev.dev ,ffmpeg.dev.doc ,ffmpeg.dev.lib ,ffmpeg.dev.man)</li>
    <li>ffmpeg.doc (ffmpeg.doc.bin ,ffmpeg.doc.data ,ffmpeg.doc.dev ,ffmpeg.doc.doc ,ffmpeg.doc.lib ,ffmpeg.doc.man)</li>
    <li>ffmpeg.lib (ffmpeg.lib.bin ,ffmpeg.lib.data ,ffmpeg.lib.dev ,ffmpeg.lib.doc ,ffmpeg.lib.lib ,ffmpeg.lib.man)</li>
    <li>ffmpeg.man (ffmpeg.man.bin ,ffmpeg.man.data ,ffmpeg.man.dev ,ffmpeg.man.doc ,ffmpeg.man.lib ,ffmpeg.man.man)</li>
  </ul>
</details>

**Comment by vcunat** [comment]
I think this would be nice to get into PR #319254.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 318912` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages failed to build:</summary>
  <ul>
    <li>ffmpeg</li>
    <li>ffmpeg.bin (ffmpeg.bin.bin ,ffmpeg.bin.data ,ffmpeg.bin.dev ,ffmpeg.bin.doc ,ffmpeg.bin.lib ,ffmpeg.bin.man)</li>
    <li>ffmpeg.data (ffmpeg.data.bin ,ffmpeg.data.data ,ffmpeg.data.dev ,ffmpeg.data.doc ,ffmpeg.data.lib ,ffmpeg.data.man)</li>
    <li>ffmpeg.dev (ffmpeg.dev.bin ,ffmpeg.dev.data ,ffmpeg.dev.dev ,ffmpeg.dev.doc ,ffmpeg.dev.lib ,ffmpeg.dev.man)</li>
    <li>ffmpeg.doc (ffmpeg.doc.bin ,ffmpeg.doc.data ,ffmpeg.doc.dev ,ffmpeg.doc.doc ,ffmpeg.doc.lib ,ffmpeg.doc.man)</li>
    <li>ffmpeg.lib (ffmpeg.lib.bin ,ffmpeg.lib.data ,ffmpeg.lib.dev ,ffmpeg.lib.doc ,ffmpeg.lib.lib ,ffmpeg.lib.man)</li>
    <li>ffmpeg.man (ffmpeg.man.bin ,ffmpeg.man.data ,ffmpeg.man.dev ,ffmpeg.man.doc ,ffmpeg.man.lib ,ffmpeg.man.man)</li>
  </ul>
</details>

**Comment by Atemu** [comment]
Wait, it says *failed* to build! :o

That's super easy to miss.

`ffmpeg` is building on the PR head for me though. Which "ffmpeg" is this?

**Comment by risicle** [comment]
@siraben do you have logs?

My test builds for this were cherry-picked to staging-next (so they would take less than a day)

**Comment by siraben** [comment]
Bah, I wish nixpkgs-review kept logs. I am rerunning the build now.

**Comment by risicle** [comment]
`nixpkgs-review` doesn't, but `nix log` does if you know the output path (or `.drv` path) that was built.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 318912` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>ffmpeg</li>
    <li>ffmpeg.bin (ffmpeg.bin.bin ,ffmpeg.bin.data ,ffmpeg.bin.dev ,ffmpeg.bin.doc ,ffmpeg.bin.lib ,ffmpeg.bin.man)</li>
    <li>ffmpeg.data (ffmpeg.data.bin ,ffmpeg.data.data ,ffmpeg.data.dev ,ffmpeg.data.doc ,ffmpeg.data.lib ,ffmpeg.data.man)</li>
    <li>ffmpeg.dev (ffmpeg.dev.bin ,ffmpeg.dev.data ,ffmpeg.dev.dev ,ffmpeg.dev.doc ,ffmpeg.dev.lib ,ffmpeg.dev.man)</li>
    <li>ffmpeg.doc (ffmpeg.doc.bin ,ffmpeg.doc.data ,ffmpeg.doc.dev ,ffmpeg.doc.doc ,ffmpeg.doc.lib ,ffmpeg.doc.man)</li>
    <li>ffmpeg.lib (ffmpeg.lib.bin ,ffmpeg.lib.data ,ffmpeg.lib.dev ,ffmpeg.lib.doc ,ffmpeg.lib.lib ,ffmpeg.lib.man)</li>
    <li>ffmpeg.man (ffmpeg.man.bin ,ffmpeg.man.data ,ffmpeg.man.dev ,ffmpeg.man.doc ,ffmpeg.man.lib ,ffmpeg.man.man)</li>
  </ul>
</details>

**Comment by siraben** [comment]
Ok, everything builds. Sorry for the noise. I found out some of the RAM sticks in my builder were faulty... 😆

**Comment by vcunat** [comment]
I think I did test the builds as well, though I'm not sure what I based the commits on.

**Pull request merged**